src/rpc/message.cpp

```diff
@@ -132,30 +93,13 @@ FullMessage::FullMessage(const std::string& json_string, bool request)
   }
 }
 
-std::string FullMessage::getJson()
-{
-
-  if (!doc.HasMember(id_field))
-  {
-    doc.AddMember(id_field, rapidjson::Value("unused"), doc.GetAllocator());
-  }
-
-  rapidjson::StringBuffer buf;
-
-  rapidjson::Writer<rapidjson::StringBuffer> writer(buf);
-
-  doc.Accept(writer);
-
-  return std::string(buf.GetString(), buf.GetSize());
-}
-
 std::string FullMessage::getRequestType() const
 {
   OBJECT_HAS_MEMBER_OR_THROW(doc, method_field)
```

Comment:
Three things here:

- We check whether we have the member we are looking for, but we do not check that it is the right type. If the member exists but is, say, an integer or an array, you could still get a RapidJSON assertion (terminating the program).
- `GetString` returns a, requiring a `strlen` for the size to be correctly determined. We should use `GetStringLength` here instead (RapidJSON already knows the size).
- We could avoid copying (and possible allocation, depending on SSO and method name length) by using a `boost::string_view` instead of an `std::string`

Reply:
The constructor for the object throws if the root type isn't object. I wanted to do away with the temporary `std::string` construction, but was kind of pushing this for some future PR. I can reorganize this a bit now ... ?

Reply:
I think I may not have explained myself clearly enough about checking the type. `doc` is obviously an object, that isn't the issue. The issue is that we check whether the object has the required key, but not whether the value for that key is of the right type.

Assuming something like the following JSON:

```
{
    "method": 100
}
```

The code would check whether the JSON value is an object (it is) and then whether it has the key `method` (it does). It then tries to convert the value for key `method` to a string, which fails because it's a number instead.

I'm fine with postponing the elimination of the temporary `std::string`, since the PR is already somewhat largish.

Reply:
I missed this while updating the code. This needs to be a separate PR so it can be backported easier.

Reply:
See PR #6361 .